dataset/bin.cpp

```diff
@@ -444,7 +451,10 @@ DataArray bin(const DataArrayConstView &array,
     end.values<scipp::index>().as_span().back() = array.dims()[dim];
     const auto indices = zip(begin, end);
     const auto tmp = make_non_owning_bins(indices, dim, array);
-    auto target_bins_buffer = makeVariable<scipp::index>(array.dims());
+    auto target_bins_buffer =
+        (array.dims().volume() > std::numeric_limits<int32_t>::max())
+            ? makeVariable<int64_t>(array.dims())
```

Comment:
👍  should this size check additionally be introduced into `makeVariable` to prevent undefined behaviour - can just throw?

Reply:
You mean if someone attempts to pass `int64_t` data when creating a variable of type `int32_t`?

Reply:
On further consideration, I think the problem I was suggesting only relates to and edge-case regarding the production of indices,  and is not general to `makeVariable`. Leave as is.